File: Problems/No.111_Minimum_Depth_of_Binary_Tree/minimum_depth_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        
        # parameters
        Q = []
        depth = 0
        
        # start compute, using BFS method
        if not root:
            return 0
        else:
            Q.append(root)
            while (len(Q)):
                Q_length = len(Q)
                for i in range(Q_length):
                    temp = Q[0]
                    del Q[0]
                    
                    if (temp.left==None and temp.right==None):
                        return depth + 1
                    else:
                        if (temp.left):
                            Q.append(temp.left)
                        if (temp.right):
                            Q.append(temp.right)
                depth += 1
            return depth
            

File: Problems/No.111_Minimum_Depth_of_Binary_Tree/minimum_depth_of_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        
        // parameters
        int left, right;
        
        // start compute
        if (root == NULL) return 0;
        else {
            
            left = minDepth(root->left);
            right = minDepth(root->right);
            return (!min(left, right)) ? max(left, right)+1 : min(left, right)+1;
            
        }
        
    }
};
